### DISCUSSION TITLE
Performance Suggestions / Best Practices

### DISCUSSION THREAD

**Discussion by dkellenb** [mapstruct/mapstruct#3274] [Q&A]
Mapstruct is great regarding mapping speed at runtime, as the code is generated at build time. We have a lot of mappers in our codebase > 50, most of them are using other (shared) mappers. Unfortunately, to build becomes slower and slower due to the many mapstruct mappers to be generated.

- Are the best practices how to organize various mappers?
- Is one or some small number of shared mappers better than many smaller mappers explicitly added? (examples for shared mappers: mapping Primitives to Id-Objects: Long => CustomerId, String <=> LocalDate/OffsetDateTime/...., XmlGregorianCalendar <=> LocalDateTime/OffsetDateTime, ...)
- Should we favor multiple mappers (for each aggregate / domain object) or few mappers (one for REST, one for each REST client)

It would be nice to see the best practices documented either here or on the official website.

**Comment by filiphr** [comment]
The amount of mappers you have should not impact the build performance. However, the amount of shared mappers / methods that can be reused is something that impacts the performance, since we are checking each one if it can map between the types.

I think that the biggest impact on performance has the complex / transitive mapping. This means that we are trying to find if there is a mapping between different types by using an intermediate type. e.g. to map A -> C we are going to see if there is something that can be mapped from A -> B -> C. You can try doing `@Mapper(mappingControl = NoComplexMapping.class)` and see how it impacts your performance.

In addition to that, if it is possible for you you can share your mappers privately with me and / or you can run a Profiler during the build time and share that privately with me. This would allow us to find bottle necks and perhaps fix them. We had a similar problem before with Kotlin.

**Reply by dkellenb** [reply]
Thank you @filiphr  for your answer. What is the easiest way to profile Mapstruct during build time? Are there some options which can be activated for easy analysis?

Our `mvn clean compile` step takes 194889ms, Mapstruct alone takes 146103ms:

![image](https://github.com/mapstruct/mapstruct/assets/9881798/924e279d-ce33-46eb-a06f-ed5ff8a9a8c7)

**Reply by filiphr** [reply]
>Mapstruct alone takes 146103ms:

That is a lot of time. 

>What is the easiest way to profile Mapstruct during build time? Are there some options which can be activated for easy analysis?

Unfortunately, we do not have anything specific. However, any standard Java profiler will work. You should be able to set it up by setting up one on the maven process itself. 

How did you come up with this image? That is already giving some information. From what I can see if you disable complex mapping (2 step mapping) you are going to shave off 133 seconds. Basically, the `MethodMethod#getBestMatch` is used for doing different permutations to see if something can be mapped by using an intermediate method. If you can do `@Mapper(mappingControl = NoComplexMapping.class)` on your mappers or on your mapper config then you would disable that path.

I know that adding it to every mapper or config might be a bit cumbersome, but that is the only way for now. If you are interested in a different approach for this, I would suggest to raise an issue to improve this a bit. Perhaps we even should disable 2-step mapping by default.

**Reply by dkellenb** [reply]
Hi @filiphr

Thank you for the fast reply. I have used VisualVM (https://visualvm.github.io/index.html) which came in the past with each JDK.

We are mapping many WebService classes, where properties and objects are encapsulated in a `JAXBElement<T>` (transforming from `TargetClass map(JAXBElement<SourceClass> source)`, therefore mapping mostly contains of two steps: First extracting the value from `JAXBElement<>` and second apply the mapping from `SouceClass` to `TargetClass`. I'll try, if `NoComplexMapping.class` would be beneficial to us, without need to add too many mapping definitions.

**Reply by filiphr** [reply]
Perhaps in your case you really need the 2-step mapping. However, I do see that we could try and solve that in a similar way as the WIP PR https://github.com/mapstruct/mapstruct/pull/3183 for optional

**Reply by dkellenb** [reply]
Yes, it looks absolutely similar. But mapping in the opposite direction needs an ObjectMapper, where we have again a two step mapping:
`JAXBElement<TargetClass> map(SourceClass source)`:

1. `TargetClass map(SourceClass source)` (this one typically will be specified by the developer)
2.  `JAXBElement<TargetClass> map(TargetClass targetClass)` (generated `ObjectMapper` by JAXB)

IMO the opposite direction is not similar to the Optional case, because there are different constructors.

**Reply by filiphr** [reply]
If we find a way to treat `JAXBElement<T>` as some kind of a value / container holder then we can apply special type of mapping for those. Right now the 2 step mapping does a variation of all possible mappings to be able to find a single mapping, if it finds more it doesn't use it. 

So if there needs to be a mapping from `SourceClass` -> `JAXBElement<TargetClass>` (i.e. `SourceValue` -> `ValueHolder<TargetValue>`) the steps that we should do should be:

* Look for a user defined mapping `JAXBElement<TargetClass> map(SourceClass source)`
* Look for user defined mapping `TargetClass map(SourceClass source)` and look for user defined mapping `JAXBElement<TargetClass> map(TargetClass targetClass)` 

When I say user define I also mean one generated by an `ObjectMapper` by JAXB (since this `ObjectMapper` is then given to MapStruct via `Mapper#uses`)

Basically we need to be able to mark certain generic classes as value holders. Really similar to what we are already doing for collection mappings.
